File: frontend/lib/features/insight/data/weekly_report.dart
class WeeklyTakeaway {
  final String title;
  final String body;
  const WeeklyTakeaway({required this.title, required this.body});

  factory WeeklyTakeaway.fromJson(Map<String, dynamic> json) =>
      WeeklyTakeaway(title: json['title'] as String, body: json['body'] as String);
}

class WeeklyReport {
  final DateTime weekStart;
  final String headline;
  final String bodyMd;
  final List<WeeklyTakeaway> takeaways;
  final DateTime generatedAt;

  const WeeklyReport({
    required this.weekStart,
    required this.headline,
    required this.bodyMd,
    required this.takeaways,
    required this.generatedAt,
  });

  factory WeeklyReport.fromJson(Map<String, dynamic> json) => WeeklyReport(
        weekStart: DateTime.parse(json['week_start'] as String),
        headline: json['headline'] as String,
        bodyMd: json['body_md'] as String,
        takeaways: (json['takeaways'] as List<dynamic>? ?? const [])
            .map((e) => WeeklyTakeaway.fromJson(e as Map<String, dynamic>))
            .toList(),
        generatedAt: DateTime.parse(json['generated_at'] as String),
      );
}

File: backend/app/models/weekly_report.py
"""WeeklyReport — LLM-generated 7-day narrative report."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import Date, DateTime, ForeignKey, Text, UniqueConstraint, func
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.user import User


class WeeklyReport(Base):
    __tablename__ = "weekly_reports"
    __table_args__ = (
        UniqueConstraint("user_id", "week_start", name="weekly_reports_user_week_unique"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False,
    )
    week_start: Mapped[date] = mapped_column(Date, nullable=False)
    headline: Mapped[str] = mapped_column(Text, nullable=False)
    body_md: Mapped[str] = mapped_column(Text, nullable=False)
    takeaways: Mapped[list[dict[str, Any]]] = mapped_column(JSONB, nullable=False, default=list)
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    user: Mapped[User] = relationship("User")
